consensus/context_deserialize_derive/src/lib.rs

```diff
@@ -0,0 +1,118 @@
+extern crate proc_macro;
+extern crate quote;
+extern crate syn;
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{
+    parse_macro_input, AttributeArgs, DeriveInput, GenericParam, LifetimeDef, Meta, NestedMeta,
+    WhereClause,
+};
+
+#[proc_macro_attribute]
+pub fn context_deserialize(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(attr as AttributeArgs);
+    let input = parse_macro_input!(item as DeriveInput);
+    let ident = &input.ident;
+
+    let mut ctx_types = Vec::new();
+    let mut explicit_where: Option<WhereClause> = None;
+
+    for meta in args {
+        match meta {
+            NestedMeta::Meta(Meta::Path(p)) => {
+                ctx_types.push(p);
+            }
+            NestedMeta::Meta(Meta::NameValue(nv)) if nv.path.is_ident("bound") => {
+                if let syn::Lit::Str(lit_str) = &nv.lit {
+                    let where_string = format!("where {}", lit_str.value());
+                    match syn::parse_str::<WhereClause>(&where_string) {
+                        Ok(where_clause) => {
+                            explicit_where = Some(where_clause);
+                        }
+                        Err(err) => {
+                            return syn::Error::new_spanned(
+                                lit_str,
+                                format!("Invalid where clause '{}': {}", lit_str.value(), err),
+                            )
+                            .to_compile_error()
+                            .into();
+                        }
+                    }
+                } else {
+                    return syn::Error::new_spanned(
+                        &nv,
+                        "Expected a string literal for `bound` value",
+                    )
+                    .to_compile_error()
+                    .into();
+                }
+            }
+            _ => {
+                return syn::Error::new_spanned(
+                    &meta,
+                    "Expected paths or `bound = \"...\"` in #[context_deserialize(...)]",
+                )
+                .to_compile_error()
+                .into();
+            }
+        }
+    }
+
+    if ctx_types.is_empty() {
+        return quote! {
+            compile_error!("Usage: #[context_deserialize(Type1, Type2, ..., bound = \"...\")]");
```

Comment:
It looks like we've got a lot of machinery in this proc macro that we aren't using/testing (using more than 1 context type, and using `bound`). Did you write this code by hand or is it based on another proc macro?

I think if we want to keep these features we should add some tests that exercise these code paths.

Reply:
I've added tests in [`51a3abb`](https://github.com/sigp/lighthouse/pull/7372/commits/51a3abb06deb47d1d7054442a889d85d9a58401b). Let me know if this is good :)